### ISSUE TITLE
P3 gap: decode graph replays W sequential forwards instead of one B-wide model traversal

### ISSUE THREAD

**Issue by gpillon** [gpillon/ignis#111]
Phase 3 specified one real B-wide decode traversal, but the leaf and each exact-width CUDA graph currently execute W complete per-lane forwards sequentially and batch only the final sampling operation. This is distinct from prefill/decode interleaving and from the phase-6 overlap work.

**Spec:** `.scratch/runtime/specs/03-serving-loop.md` (solution and requirement 17). Related: #64, #99, #102, #110.

**Blocked by:** —

**Comment by gpillon** [comment]
## Diagnostic evidence

### Problem

The scheduler and ABI correctly collect all decode-ready sequences into one call, and sampling is performed once over the round. The expensive model forward is not batched, however: the implementation traverses the complete model separately for lane 0, then lane 1, and so on. The exact-width CUDA graph records that same sequential operation sequence.

In other words, the current shape is:

```text
for lane in 0..B:
    run all 64 layers for lane
sample B logits columns together
```

The required shape is:

```text
run all 64 layers once over tensors/state with batch dimension B
sample B logits columns together
```

This is a specification-compliance and performance defect, not merely a naming problem. Spec 03 says that decode becomes a real batched round and requires eight lanes to stream model weights once rather than eight times.

### Direct evidence

- `kernel/src/decode_graph.cu:105-109` describes each width-W graph as W sequential lane forward passes followed by one batched sample.
- `kernel/src/step.cu:717-723` says each lane receives one sequential per-sequence traversal and explicitly says that the layer bodies are not fused across sequences.
- `.scratch/runtime/specs/03-serving-loop.md:78-83` specifies a real batched round.
- `.scratch/runtime/specs/03-serving-loop.md:112` requires one decode call spanning every ready lane so that eight lanes stream the model weights once rather than eight times.
- #99 delivered per-sequence sampling and stable buffers, but did not own a batched model traversal.
- #102 captured the existing eager sequence into exact-width graphs; graph capture removes launch overhead but does not turn sequential forwards into a tensor batch.

### Not the same as interleaving or phase-6 overlap

Interleaving is already present: the scheduler alternates a prefill chunk with a decode round so a long prefill does not monopolize the model stream. It says nothing about whether the sequences inside that decode round traverse the model together.

Phase 6 proposes prefill/decode overlap on separate streams. That can run prefill and decode concurrently, but it does not convert W sequential decode forwards into one B-wide forward. No current G4, G5, phase-6, or investigation ticket owns this missing decode implementation. #92 concerns prefill synchronization and possible B>1 prefill, which is a different path.

### How to reproduce

#### Static proof, no GPU required

From the repository root:

```powershell
rg -n -C 4 "sequential lane forward|sequential per-sequence|does not fuse" kernel/src/decode_graph.cu kernel/src/step.cu
rg -n -C 3 "real batched round|stream the model.s weights once" .scratch/runtime/specs/03-serving-loop.md
```

The first command shows the implemented sequential traversal. The second shows the contradictory requirement.

#### Runtime/profile confirmation

1. Start the CUDA server with decode graphs enabled and enough capacity for four resident decode lanes.
2. Run the existing `ignis-bench g3` C=4 workload, or submit four equal 8,192-token prompt / 256-token-cap requests concurrently.
3. Capture a decode round with Nsight Systems or equivalent CUDA tracing.
4. Inspect one width-4 round: before the single batched sampling operation, four complete model-forward kernel sequences appear in lane order.
5. Repeat at width 1. Width 4 increases the number of forward traversals rather than changing each layer invocation to operate over B=4.

The existing exact-width graph correctness test can still pass because it checks eager/replay token equivalence, not whether the eager path is genuinely batched. Likewise, a passing C=4 throughput comparison does not prove that weights are streamed once per round.

### Consequences

- Model weights and layer work are revisited once per active lane instead of being shared by the round.
- Decode-round latency grows with B much more than intended.
- CUDA graphs hide launch overhead but preserve the wrong computational shape.
- G3 was closed while a load-bearing requirement from its master/spec remained unimplemented.
- Deferring #110 solely to phase-6 overlap would conflate two independent problems.

### Implementation seam

Keep the public `ignis_program_decode` interface and per-sequence state semantics. Replace the leaf internals with a B-wide traversal for exact widths 1..8: each layer consumes batch-shaped activations and the corresponding sequence state/page tables, producing `[vocab, B]` logits for the existing batched sampler. Exact-width graphs should capture that B-wide traversal, never pad lanes, and retain the eager fallback.

### Acceptance criteria

- [ ] Widths 2..8 execute one B-wide model traversal, not a host loop of B calls to the batch-1 forward.
- [ ] Per-sequence KV, GDN, conv, position, RNG, penalties and sampling parameters remain isolated and correctly indexed.
- [ ] Exact widths 1..8 remain unpadded; eager fallback remains available.
- [ ] Eager and CUDA-graph paths produce identical tokens at every width 1..8.
- [ ] A GPU trace or explicit leaf instrumentation proves one layer traversal per decode round at B>1 rather than B complete layer traversals.
- [ ] C=1 has no material regression and C=4 is remeasured with the G3 instrument.
- [ ] Teacher-forced canary and G2 self-oracle remain green.
- [ ] Leaf GPU tests and `cargo test --workspace` pass.

### Relationship to #110

This defect is a credible contributor to excessive decode-round cost and therefore to ITL, but it should not pre-judge the separate measurement and prefill-boundary findings in #110. Resolve and measure this independently; then reassess the remaining ITL gap.

**Comment by gpillon** [comment]
## Runtime confirmation, and this owns #110's ITL gap

The static evidence above is now backed by a number, taken off #110's live/live records (`.scratch/g3-gate-110/`, session `g3-110-window-cancel-v3`) without needing a new run.

A decode lane's inter-token interval that intersects **no** prefiller window is a decode round on its own. Both records carry thousands of them.

| | ignis | reference | ratio |
|---|---:|---:|---|
| decode round, B=4 (median, outside every window) | 70.2 ms | 17.3 ms | 4.06 |
| decode round, B=1 (from the C=1 cell) | 14.6 ms | 16.1 ms | 0.91 |
| B=4 cost relative to own B=1 | **4.81x** | 1.07x | |

At B=1 ignis is ahead of the reference. At B=4 it pays 4.81x its own single-lane round, while the reference pays 1.07x. That is exactly the shape this issue describes: the round's cost scales with B because the model is traversed B times, and only the sampling is shared. The reference's 1.07x is what a real B-wide traversal looks like.

## It is the whole of #110's ITL p95 gap

Decomposing #110's blocked intervals with the round now known:

| term | ignis | reference | ratio |
|---|---:|---:|---|
| prefill chunk, 1,024 tokens at 32K context | 116.3 ms | 141.2 ms | 0.82 |
| decode round, B=4 | 70.2 ms | 17.3 ms | 4.06 |
| blocked ITL interval, the sum | 186.5 ms | 158.5 ms | 1.18 |

The sums match the measured means of 180.8 ms and 155.4 ms. ignis's prefill is 21% *faster*; the entire gap is this defect. Bringing the round to the reference's order puts the blocked interval near 136 ms against 158 ms, a ratio of about 0.86, which clears the G3 gate with margin.

#110 was closed as deferred to phase 4 on the opposite reading and has been reopened, blocked by this issue. Your closing note above was right that this should not pre-judge #110's measurement findings; it turned out to subsume them.

## One addition to the acceptance criteria

Criterion "a GPU trace or explicit leaf instrumentation proves one layer traversal per decode round at B>1" can also be checked without a profiler, from the G3 instrument alone: a real B-wide traversal shows B=4 costing near 1.0x the B=1 round. The reference's 1.07x is the target shape, and ignis's current 4.81x is the baseline to beat.

**Comment by gpillon** [comment]
## The round is one batch-wide traversal now (ADR 0020) — landed, not closed

Commits `707c24e` and `01f96fe` on `main`.

`ignis_decode_graph_run_batch` (`kernel/src/decode_graph.cu`) runs a whole
decode round as a single traversal of the model at the round's exact width:
embedding, every decoder layer's graph-safe body once with batch-shaped
activations, final norm, output head writing `[vocab, width]` straight into
the sampler's staging buffer. The per-lane loop and its per-lane logits
column copy are gone.

The layer bodies take a **width** instead of a lane index. GQA runs A1 with
the lanes on its batch axis at W=1 per row, `kv_table_rows` reading the whole
staged slot vector; GDN runs `causal_conv1d_silu_snapshot` and
`gated_delta_net_snapshot` the same way, with `initial_state_slots ==
snapshot_base_slots == decode_graph_slots`. No vendored kernel changed: the
ops already admitted B=2..8 at W=1 for exactly this case, which is what ADR
0019's last Consequence anticipated.

Two smaller things fell out. RoPE positions are read straight from
`sampling_decode_positions` — the lanes' positions are unrelated, so there is
no base to offset from — which retires `decode_graph_zero` and the
`offset_i32_positions` call. And `decode_graph_scratch` is sized once for one
token per lane across `IGNIS_DECODE_MAX_BATCH` lanes, shared by every width.

### One traversal serves both paths

Graph capture records that traversal; a round at a width whose capture failed
enqueues the identical call sequence directly. The eager fallback is still a
fallback — a capture failure degrades performance and never refuses service —
but it is no longer a second, differently shaped implementation of the round.

That is also the fix for a latent defect this uncovered: #102's graph path
never advanced `seq->gqa_positions`, because it read its positions from
device staging, so a round that fell back to eager after a replay read a
stale counter. `ignis_program_decode` now advances it once per round, for
whichever path ran.

### The profiler-free check this issue asked for

`crates/core/tests/decode_round_scaling_gpu.rs` measures it and fails above
2.0x. On a free 5090, greedy, 256-token context, debug build:

| | ignis |
|---|---:|
| decode round, B=1 | 12.31 ms |
| decode round, B=4 | 14.06 ms |
| B=4 relative to own B=1 | **1.14x** |

Against 4.81x before, and the reference's 1.07x. Treat it as a shape check,
not a gate reading: it is a synthetic micro-benchmark at short context, not
the G3 instrument at 32K.

### Acceptance criteria

- [x] **Widths 2..8 execute one B-wide model traversal.**
  `ignis_program_stats`'s `kernel_count` for a decode round is now the layer
  count at every width and on both paths, where it was the layer count times
  the width. Asserted on every round of the width 1..8 loop in
  `crates/core/tests/decode_graph_gpu.rs`.
- [x] **Per-sequence KV, GDN, conv, position, RNG, penalties and sampling
  parameters remain isolated and correctly indexed.** Every one of them is
  read from device staging indexed by row. New coverage presents the same
  four sequences to the round in reverse row order and requires each one's
  token stream to be unchanged. That test is greedy throughout, so it does
  not yet stress the RNG and penalty halves — filed as #115, test-side only.
- [x] Exact widths 1..8 remain unpadded; eager fallback remains available.
- [x] Eager and CUDA-graph paths produce identical tokens at every width 1..8.
- [x] Leaf instrumentation proves one layer traversal per decode round at
  B>1 — the `kernel_count` assertion above, plus the ratio.
- [ ] **C=1 has no material regression and C=4 is remeasured with the G3
  instrument.** Not done. Needs a live/live session against the reference
  (ADR 0015) and the card. See below.
- [x] Teacher-forced canary and G2 self-oracle remain green —
  `teacher_forced_canary_agreement_meets_the_g1_floor` and
  `chunked_and_per_token_prefill_agree_on_a_long_prompt`, both ok in the
  profile run.
- [x] Leaf GPU tests and `cargo test --workspace` pass. Full profile green:
  33/33 kernel op tests, 69 Rust GPU test blocks ok, 0 failed.

The GPU profile also needed a one-line unrelated fix to run at all:
`crates/runtime/src/cuda_leaf.rs`'s sampling-mapping test had not been
updated when `DecodeParams` gained `ignore_eos` (`59a0511`), so the
`--features cuda` build had been failing to compile since then.

### Why this stays open

Seven criteria of eight. The eighth is not a formality: this change also
altered the B=1 path, which now goes through the batched code with a
conservative envelope and the pool-wide block table instead of the
single-sequence view. The C=1 cell carries roughly 0.13 ms of margin on a
13.2 ms step — exactly where a regression hides, and exactly what the
synthetic ratio above cannot rule out.

The measurement needed is the same live/live run #110 needs for its ITL
cell. Whether it is taken under this issue or under #110 is a scheduling
call; either way it is one run, not two.

**Comment by gpillon** [comment]
## The last criterion, closed: no C=1 regression, C=4 remeasured

Pooling every C=1/C=4 sample taken across this work (the original indicative
A/B, plus #116's two-launch investigation, which happened to also be on this
tree):

| | pre-#111 | post-#111 |
|---|---|---|
| C=1, tok/s | 63.4–73.9 (1 launch, 5 samples) | 63.9–73.4 (3 launches, 10 samples) |
| C=4, tok/s | 17.9–20.9 (1 launch, 5 samples) | 43.1–48.3 (3 launches, 10 samples) |

C=1: the post range sits entirely inside the pre range. No regression signal
— given what #116 taught us about launch-to-launch spread, a range this wide
already covers what more launches would show. C=4: the two ranges don't come
close to overlapping — more than double, robust to the launch variance
because the gap dwarfs it.

This isn't the G3 gate verdict (that needs a live/live reference leg, ADR
0015, and is #110's job), but it's what this criterion actually asked for:
ignis's own before/after. Closing on it.

All eight acceptance criteria are now met. Closing.
